reco_utils/recommender/surprise/surprise_utils.py

```diff
@@ -0,0 +1,49 @@
+import numpy as np
+import pandas as pd
+
+from reco_utils.common.constants import DEFAULT_USER_COL, DEFAULT_ITEM_COL, DEFAULT_RATING_COL
+
+
+def compute_predictions(algo, data, usercol=DEFAULT_USER_COL, itemcol=DEFAULT_ITEM_COL):
+    """
+    Computes predictions of an algorithm from Surprise on the data
+    Args:
+        algo (surprise.prediction_algorithms.algo_base.AlgoBase): an algorithm from Surprise
+        data (pd.DataFrame): the data on which to predict
+        usercol (str): name of the user column
+        itemcol (str): name of the item column
+    Returns:
+        pd.DataFrame: dataframe with usercol, itemcol, prediction
+    """
+    predictions = [algo.predict(row[usercol], row[itemcol]) for (_, row) in data.iterrows()]
+    predictions = pd.DataFrame(predictions)
+    predictions = predictions.rename(index=str, columns={'uid': usercol, 'iid': itemcol, 'est': 'prediction'})
+    return predictions.drop(['details', 'r_ui'], axis='columns')
+
+
+def compute_all_predictions(algo, data, usercol=DEFAULT_USER_COL, itemcol=DEFAULT_ITEM_COL, recommend_seen=False):
+    """
+    Computes predictions of an algorithm from Surprise on all users and items in data.
+    Args:
+        algo (surprise.prediction_algorithms.algo_base.AlgoBase): an algorithm from Surprise
+        data (pd.DataFrame): the data from which to get the users and items
+        usercol (str): name of the user column
+        itemcol (str): name of the item column
+        recommend_seen (bool): flag to include (user, item) pairs that appear in data
+    Returns:
+        pd.DataFrame: dataframe with usercol, itemcol, prediction
+    """
+    preds_lst = []
+    for user in data[usercol].unique():
+        for item in data[itemcol].unique():
+            preds_lst.append([user, item, algo.predict(user, item).est])
+
+    all_predictions = pd.DataFrame(data=preds_lst, columns=[usercol, itemcol, "prediction"])
```

Comment:
we might want to create a function that returns the top k, based on https://github.com/Microsoft/Recommenders/blob/master/reco_utils/recommender/sar/sar_singlenode.py#L290. This will reduce the memory consumption. What do you think?

Reply:
We need to see this more broadly in a new issue. Because we also have https://github.com/Microsoft/Recommenders/blob/995b0789d449c6d485e76fe01e387e4148b281e4/reco_utils/evaluation/python_evaluation.py#L594
So, what happens with SAR is that the k largest computation happens twice.

Reply:
there is an issue opened by @gramhagen https://github.com/Microsoft/Recommenders/issues/505 I will add there our comments and the reference made by Max on topk in spark